electrum/lnchannel.py

```diff
@@ -1091,32 +1095,71 @@ def _assert_can_add_htlc(self, *, htlc_proposer: HTLCOwner, amount_msat: int,
         if not ignore_min_htlc_value:
             if amount_msat <= 0:
                 raise PaymentFailure("HTLC value must be positive")
-            if amount_msat < chan_config.htlc_minimum_msat:
+            if amount_msat < chan_payment_constraints.htlc_minimum_msat:
                 raise PaymentFailure(f'HTLC value too small: {amount_msat} msat')
 
         # check proposer can afford htlc
         max_can_send_msat = self.available_to_spend(htlc_proposer, strict=strict)
         if max_can_send_msat < amount_msat:
             raise PaymentFailure(f'Not enough balance. can send: {max_can_send_msat}, tried: {amount_msat}')
 
+        # check "max_htlc_value_in_flight_msat"
+        current_htlc_sum = htlcsum(self.hm.htlcs_by_direction(htlc_receiver, direction=RECEIVED, ctn=ctn).values())
+        if current_htlc_sum + amount_msat > chan_config.max_htlc_value_in_flight_msat:
+            raise PaymentFailure(f'HTLC value sum (sum of pending htlcs: {current_htlc_sum/1000} sat '
+                                 f'plus new htlc: {amount_msat/1000} sat) '
+                                 f'would exceed max allowed: {chan_config.max_htlc_value_in_flight_msat/1000} sat')
+
+    def determine_payment_constraints(self, htlc_proposer: HTLCOwner) -> ChannelPaymentConstraints:
+        """Raises PaymentFailure if no valid amount range can be determined, or if no HTLCs can be added.
+        (this is relevant both for forwarding and endpoint)
+        """
+        htlc_receiver = htlc_proposer.inverted()
+        ctn = self.get_next_ctn(htlc_receiver)
+        chan_config = self.config[htlc_receiver]
+        if self.get_state() != ChannelState.OPEN:
+            raise PaymentFailure('Channel not open', self.get_state())
+        if htlc_proposer == LOCAL:
+            if not self.can_send_ctx_updates():
+                raise PaymentFailure('Channel cannot send ctx updates')
+            if not self.can_send_update_add_htlc():
+                raise PaymentFailure('Channel cannot add htlc')
```

Comment:
I would not raise `PaymentFailure` inside `determine_payment_constraints`, because that method is called by `num_sats_can_send/receive`. (those `PaymentFailures` are triggered by network conditions, while `num_sat_can_send/receive` should return stable results). It also simplifies  the upstream code if you do not have to catch `PaymentFailure`

Reply:
Yes, I left this in to keep the behavior of `num_sats_can_xxx` the same as I took the constraints part out, and I need the error path using exceptions to propagate the error message out.

I think it feels more appropriate if I raise e.g. `ChannelConstraintFailure` or something, catch these in `num_sats_can_xxx`, and re-raise there with the same error message, but in a `PaymentFailed` class.

Reply:
> It also simplifies the upstream code if you do not have to catch `PaymentFailure`

Agreed. The exceptions are useful within the context of `num_sats_can_xxx`, but not so much in other places. I can add a convenience method that just returns `None`

Reply:
The way the current code works is that `get_channels_for_sending/receiving` filters the channels that cannot send due to transient network conditions. If you need to add new checks, please them where the existing checks are.